src/core/protocol-authorization.ts

```diff
@@ -155,43 +158,17 @@ export class ProtocolAuthorization {
   }
 
   /**
-   * Verifies the requester of the given message is allowed actions based on the rule set.
-   * @throws {Error} if requester not allowed.
+   * Verifies the actions specified in the given message matches the allowed actions in the rule set.
+   * @throws {Error} if action not allowed.
    */
-  private static verifyAllowedRequester(
```

Comment:
For my understanding: what's the reason that prompted the decision to merge the logic into `verifyAllowedActions()`? Was it because it's hard to understand? Or redundant code? Or?

The intention behind having two separate methods here were:
1. It is easier to know what the method is doing (in theory) without reading the implementation.
2. It gives precise error message when error does occur.

The new implementation removes both benefits above, hence want to discuss it.

Reply:
I tried keeping `verifyAllowedRequester` and `verifyAllowedActions` separate, but I found that they were becoming redundant. Plus, I felt that the implementation was more intuitive by combining them. I'm down to jam on keeping them separate if you feel that's best. Here was my thought process:

We had these lines originally:
```typescript
    if (allowRule.anyone !== undefined) {
      allowedActions = allowRule.anyone.to;
    } else if (allowRule.recipient !== undefined) {
      allowedActions = allowRule.recipient.to;
    }
    const inboundMessageAction = methodToAllowedActionMap[incomingMessageMethod];
    if (!allowedActions.includes(inboundMessageAction)) {
      throw new Error(`inbound message action '${inboundMessageAction}' not in list of allowed actions ${allowedActions}`);
    }
```

Consider an `allowRule` with both `anyone` and `recipients` defined. It's possible that `recipients` have more allowed actions than `anyone`, and that the `inboundMessageAction` is in `recipient.to`. The above code would erroneously block authorization because it would hit the `if (allowRule.anyone !== undefined)`.

We could fix this by turning the `else if` into a regular `if`, and adding allowed rules into a set of `allowedActions`. But it gets odd when you add `author`.
```typescript
    if (allowRule.anyone !== undefined) {
      allowRule.anyone.to.forEach(action => allowedActions.add(action));
    }
    if (allowRule.recipient !== undefined) {
      allowRule.recipient.to.forEach(action => allowedActions.add(action));
    }
    if (allowRule.author !== undefined) {
      allowRule.author.to.forEach(action => allowedActions.add(action));
    }
```

This is bad because we don't want to allow all the actions in all the rules. We need to check that the message is actually from an `author` or a `recipient`.

```typescript
    if (allowRule.anyone !== undefined) {
      allowRule.anyone.to.forEach(action => allowedActions.add(action));
    }
    if (allowRule.recipient !== undefined) {
      if message is from a recipient {
        allowRule.recipient.to.forEach(action => allowedActions.add(action));
      }
    }
    if (allowRule.author !== undefined) {
      if message is from an author {
        allowRule.recipient.to.forEach(action => allowedActions.add(action));
      }
    }
```
We already do those checks in `verifyAllowedRequester`, so `verifyAllowedRequester` is redundant.

Reply:
Thanks @diehuxx for the full context, I did not at all consider the possibility of having multiple "roles" under the same `allow` property! I will continue with the review with this new knowledge!